lib/ecto/changeset/relation.ex

```diff
@@ -240,11 +240,11 @@ defmodule Ecto.Changeset.Relation do
   end
 
   defp single_change(relation, new, current_pks, new_pks, fun, current) do
-    if new_pks.(new) == current_pks.(current) do
+    if new_pks.(new) == current_pks.(current) or relation.on_replace == :update do
```

Comment:
Let's pattern match `%{on_replace: on_replace} = relation` on the signature instead. :)